### PR TITLE
Modify spnego to use the winpr asn.1 library

### PR DESCRIPTION
**Pull request by fifthdegree**

### CODE REVIEW

**Comment by hardening** [comment]
@freerdp-bot can you test that good stuff please ?

File: winpr/libwinpr/sspi/Negotiate/negotiate.c [hunk 1]
```diff
@@ -260,12 +180,12 @@ static const char* negotiate_mech_name(const sspi_gss_OID_desc* oid)
 		return "Unknown mechanism";
 }
 
-static const Mech* negotiate_GetMechByOID(sspi_gss_OID_desc oid)
+static const Mech* negotiate_GetMechByOID(WinPrAsn1_OID oid)
```
Comment (hardening) [inline winpr/libwinpr/sspi/Negotiate/negotiate.c:183]: Shall be changed to `static const Mech* negotiate_GetMechByOID(const WinPrAsn1_OID* oid)` and use a local variable later

File: winpr/libwinpr/sspi/Negotiate/negotiate.c [hunk 2]
```diff
@@ -349,245 +269,209 @@ static BOOL negotiate_get_config(BOOL* kerberos, BOOL* ntlm)
...
-				else
-					return FALSE;
 				break;
-			case 0xA1:
+			case 1:
 				/* reqFlags [1] ContextFlags BIT STRING (ignored) */
-				if (tag == 0x03 && token->init)
+				if (token->init)
 				{
+					WinPrAsn1DecPeekTagAndLen(&dec2, &tag, &len);
```
Comment (hardening) [inline winpr/libwinpr/sspi/Negotiate/negotiate.c:438]: Shall check the return value

File: winpr/libwinpr/sspi/Negotiate/negotiate.c [hunk 3]
```diff
@@ -679,36 +563,50 @@ static SECURITY_STATUS SEC_ENTRY negotiate_InitializeSecurityContextW(
 
 	if (!context)
 	{
+		enc = WinPrAsn1Encoder_New(WINPR_ASN1_DER);
+		if (!enc)
+			return SEC_E_INTERNAL_ERROR;
```
Comment (hardening) [inline winpr/libwinpr/sspi/Negotiate/negotiate.c:568]: `return SEC_E_INSUFFICIENT_MEMORY`

File: winpr/libwinpr/sspi/Negotiate/negotiate.c [hunk 4]
```diff
@@ -679,36 +563,50 @@ static SECURITY_STATUS SEC_ENTRY negotiate_InitializeSecurityContextW(
...
+					cred->valid = FALSE;
+					continue;
+				}
 
-			/* If the mechanism failed we can't use it; skip */
-			if (IsSecurityStatusError(status))
-				cred->valid = FALSE;
-			else
 				init_context.mech = cred->mech;
+				status = SEC_E_INTERNAL_ERROR;
```
Comment (hardening) [inline winpr/libwinpr/sspi/Negotiate/negotiate.c:599]: We managed to retrieve a security context, and the status is set to an error, not sure to understand...

**Comment by hardening** [comment]
@fifthdegree I've pushed a few changes on your branch. I feel like `negotiate_AcceptSecurityContext()` is really big, isn't it possible to split the function ?

**Comment by fifthdegree** [comment]
> @fifthdegree I've pushed a few changes on your branch. I feel like `negotiate_AcceptSecurityContext()` is really big, isn't it possible to split the function ?

Anything you have in mind?

**Comment by hardening** [comment]
> > @fifthdegree I've pushed a few changes on your branch. I feel like `negotiate_AcceptSecurityContext()` is really big, isn't it possible to split the function ?
> 
> Anything you have in mind?

Some proposals (only proposals) that may make thing more clean:
* perhaps the cases with a raw token and the case with SP nego could be splitted, they don't have that much in common (that may help when you will add on top of that the user2user case);
* for the first packet (when `context == NULL`), perhaps the treatment that does the selection of a response token could be moved to a function (will participate to have less code);

Just some ideas, if the code in `negotiate_AcceptSecurityContext()` is smaller it's easier to check return codes, potential leaks, ...

**Comment by hardening** [comment]
@freerdp-bot can you test again please ?

File: winpr/libwinpr/sspi/Negotiate/negotiate.c [hunk 5]
```diff
@@ -876,6 +781,9 @@ static SECURITY_STATUS SEC_ENTRY negotiate_InitializeSecurityContextW(
 	if (!negotiate_write_neg_token(output_buffer, &output_token))
 		status = SEC_E_INTERNAL_ERROR;
 
+cleanup:
+	if (enc)
```
Comment (akallabeth) [inline winpr/libwinpr/sspi/Negotiate/negotiate.c:785]: No `NULL` check required.

File: winpr/libwinpr/utils/asn1/asn1.c [hunk 6]
```diff
@@ -1401,3 +1401,12 @@ size_t WinPrAsn1DecReadContextualSequence(WinPrAsn1Decoder* dec, WinPrAsn1_tagId
 	Stream_Seek(&dec->source, ret);
 	return ret;
 }
+
+void WinPrAsn1DecGetStream(WinPrAsn1Decoder* dec, wStream* s)
```
Comment (akallabeth) [inline winpr/libwinpr/utils/asn1/asn1.c:1405]: Maybe change to `wStream WinPrAsn1DecGetStream(WinPrAsn1Decoder* dec)`

**Review by akallabeth — APPROVED** [review]
LGTM,
some small (non blocking) remarks.

File: winpr/libwinpr/utils/asn1/asn1.c (hunk 6 above)
Comment (hardening) [reply winpr/libwinpr/utils/asn1/asn1.c:1405]: On my side I like it this way.

Comment (akallabeth) [reply winpr/libwinpr/utils/asn1/asn1.c:1405]: @hardening well, this is kind of a `allocator` function, so you can do `wStream s = WinPrAsn1DecGetStream(dec);` in a single line while preventing doing something like `wStream* s = NULL; WinPrAsn1DecGetStream(dec, s);`

File: winpr/libwinpr/utils/asn1/asn1.c [hunk 7]
```diff
@@ -1402,11 +1402,11 @@ size_t WinPrAsn1DecReadContextualSequence(WinPrAsn1Decoder* dec, WinPrAsn1_tagId
...
+wStream WinPrAsn1DecGetStream(WinPrAsn1Decoder* dec)
 {
+	wStream s = { 0 };
 	WINPR_ASSERT(dec);
-	WINPR_ASSERT(s);
 
-	Stream_StaticConstInit(s, Stream_Pointer(&dec->source),
+	Stream_StaticConstInit(&s, Stream_Pointer(&dec->source),
 	                       Stream_GetRemainingLength(&dec->source));
 }
```
Comment (hardening) [inline winpr/libwinpr/utils/asn1/asn1.c:1412]: missing `return s;`

**Comment by akallabeth** [comment]
@freerdp-bot test

**Pull request merged**